### PR TITLE
nixos/opensearch: Use DynamicUser and StateDirectory by default

### PR DESCRIPTION
**Pull request by talyz**
###### Motivation for this change
Use DynamicUser and StateDirectory by default, but still allow for setting `dataDir` to a custom path. This gets rid of the use of the deprecated option `PermissionsStartOnly`. Also, add the ability to customize user and group, since that could be useful with a custom `dataDir`.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by shyim** [comment]
OHA. Really thanks for your effort! ❤️

File: nixos/modules/services/search/opensearch.nix [hunk 1]
```diff
@@ -20,9 +23,9 @@ in
 {
 
   options.services.opensearch = {
-    enable = mkEnableOption (lib.mdDoc "Whether to enable OpenSearch.");
+    enable = mkEnableOption "OpenSearch";
```
Comment (mweinelt) [inline nixos/modules/services/search/opensearch.nix:26]: I think this broke the manual build.

Comment (talyz) [reply nixos/modules/services/search/opensearch.nix:26]: Oops, yeah, the `lib.mdDoc` still needs to be there.

Comment (shyim) [reply nixos/modules/services/search/opensearch.nix:26]: I will force push your commit :)

Comment (talyz) [reply nixos/modules/services/search/opensearch.nix:26]: Yeah, I tried to do it, but I guess GitHub didn't like that :shrug:

Comment (shyim) [reply nixos/modules/services/search/opensearch.nix:26]: Yep github is buggy. In the fork it looks correct but not in the PR 😂 . I try ammend + force push

Comment (talyz) [reply nixos/modules/services/search/opensearch.nix:26]: Seems to have done the trick :)

**Pull request merged**